### ISSUE TITLE
[@typescript-eslint/unbound-method] Ignore uses in boolean contexts

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#2515]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/unbound-method": "error"
  }
}
```

```TS
class Container {
    public method?() {
        return true;
    }
}

const foo = new Container();

if (foo.method) {
    foo.method;
}
```

**Expected Result**

Since this is an existence check, there should be no lint complaint about accessing a class instance method without preserving the `this` scope.

**Actual Result**

```ts
    foo.method;
 // ~~~~~~~~~~ [@typescript-eslint/unbound-method]
 //            Avoid referencing unbound methods which may cause unintentional scoping of `this`.
```

**Additional Info**

Continuing @felixfbecker's https://github.com/typescript-eslint/typescript-eslint/issues/522#issuecomment-517961987, since I couldn't find another issue filed on it _(is there one?)_.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.1.0` |
| `@typescript-eslint/parser`        | `4.1.0` |
| `TypeScript`                       | `4.0.2` |
| `ESLint`                           | `7.8.1` |
| `node`                             | `14.7.0` |

**Comment by bradzacher** [comment]
This _should_ be working fine.
We have passing tests in master for this: https://github.com/typescript-eslint/typescript-eslint/blob/ec7449d133e1bbca76a7df4e4b325e9bda547fb1/packages/eslint-plugin/tests/rules/unbound-method.test.ts#L73-L74

will require some investigation..

**Comment by glen-84** [comment]
There's no error on line 9 (existence check), the error is on line 10.

**Comment by bradzacher** [comment]
Yeah I can't repro this against master. The only error in master is on line 10 (which is valid)

**Comment by JoshuaKGoldberg** [comment]
Ok, I think I just did a dumb then. Sorry for the spam!
